### ISSUE TITLE
Integration of Sklearn Cross validation Strategies in Catalyst

### ISSUE THREAD

**Issue by Atharva-Phatak** [catalyst-team/catalyst#1221]
## 🚀 Feature Request

Sklearn provides multiple cross-validation strategies, Catalyst has a very user friendly API. I would like to add a method to catalyst which will enable user to do cross validation with catalyst.

### Proposal

Here are a list of possible solutions I can think of
1. Create a new class essentially inheriting from IRunner which accepts torch style datasets(with keys for input(x) and targets(y)) along with a cross val object, which will enable us to create folds. A problem may occur since while training on each fold we need to create a new instance of the model or maybe reinitialize it, which may need some major modifications.
2. Another approach could be we write our own cross val strategies customized for catalyst.
3. Maybe we create off the shelf PyTorch Style datasets for standard tasks like classification, regression,etc. Doing this we will know the input and target keys internally and maybe adapt sklearn cross validation strategies according to it.

### Additional context

This seems like a big task and hence I want your ideas on it, on how we can go on implementing it.

### Checklist
- [x] feature proposal description
- [ ] motivation
- [x] extra proposal context / proposal alternatives review
  

### FAQ
Please review the FAQ before submitting an issue:
- [x] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [x] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [x] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [x] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
-  [x] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

**Comment by Scitator** [comment]
Hi @Atharva-Phatak ,
thanks for your proposal, looks very interesting (and we definitely have something like that in mind).
Do you have a vision of how it should look like in the end? without final implementation, of course, but on a high level?

**Comment by Atharva-Phatak** [comment]
@Scitator I would personally design a new class since cross validation has so many things to keep track of a separate runner would be better in long term perspective.
1) What I would also love to design is a model callback that initiailizes model, everytime a new fold training start. 
2) Coming to the point of generating folds, I think creating a dict which holds all the folds indices would suffice at start but eventually we have to find a way to access inputs and targets to the dataset for fold generation. So I am not so sure how that would look like.

Overall High Level view for me would be

* New class for cross validation
* A model initialization callback for resetting the params. Same reset is applicable for optimizer, scheduler,etc
* A good way of logging each fold for later interpretation.
* A hacky way to reinit callbacks like model checkpoint etc, so that everytime a new model is created for fold we can just save it.

But please lets discuss more, I really think with brainstorming we can come with an initial solution so that we can push catalyst to new heights 😃

**Comment by Scitator** [comment]
🤔  do you have any pseudocode ideas?
something like
```python
class CrossValidationRunnerWrapper:

  def num_folds(self) -> int:
    # return the number of folds

  def datasets(self, fold: I) -> dict[Datasets]:
    # return the datasets for i-th fold

  def train_cross_validation(self):
    # do cross-validation magic :)
```
from my perspective, the cross-validation could be done very easily with the CustomRunner-like approach, for example: [CV - MNIST multistage fine-tuning](https://github.com/catalyst-team/catalyst#minimal-examples)

**Comment by Atharva-Phatak** [comment]
@Scitator I like the custom runner approach though how will you re-initialize the model for each fold and other callbacks as well?

Here is what I was thinking and of course I am open to other ideas!

1) Find a way to reinitialize the models, callbacks, etc. i.e how can we IRunner to initialize it ?
2) How do we get "X" and "y" for the datasets? Accept keys ? we have to figure this out.
3)  I was thinking of creating an internal fold dictionary that will have folds and indexes. So that we can torch.utils.data.Subset to get train and validation datasets/data loaders.

But please I do want to see it integrated in catalyst because catalyst has almost everything ! 🥳 

PS: I started my open source journey started with first contribution to catalyst. So thanks for accepting my first PR 😄

**Comment by Scitator** [comment]
Could you please provide some minimalistic vision on MNIST?

**Comment by Atharva-Phatak** [comment]
Yes. I was a little busy. I will create a minimal working example ASAP.
